### PR TITLE
SCRUM-3952: change vocabularyterm to ontologyterm after LinkML change

### PR DESCRIPTION
**Pull request by abecerra**

### CODE REVIEW

**Review by markquintontulloch — COMMENTED** [review]
The AnatomicalSite entity class needs the `@AGRCurationSchemaVersion` annotation updated to a min of 2.7.0 - you will also need to increase the `LinkMLSchemaConstants.LATEST_RELEASE` constant to this value

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 1]
```diff
@@ -118,7 +104,7 @@ public GeneExpressionAnnotation validateAnnotation(GeneExpressionFmsDTO geneExpr
 
 		ObjectResponse<AnatomicalSite> anatomicalSiteObjectResponse = validateAnatomicalSite(geneExpressionFmsDTO);
 		if (anatomicalSiteObjectResponse.hasErrors()) {
-			response.addErrorMessage("whereExpressed", anatomicalSiteObjectResponse.errorMessagesString());
+			response.addErrorMessage("expressionPattern", anatomicalSiteObjectResponse.errorMessagesString());
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java:107]: This change doesn't make sense to me

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 2]
```diff
@@ -127,7 +113,7 @@ public GeneExpressionAnnotation validateAnnotation(GeneExpressionFmsDTO geneExpr
 
 		ObjectResponse<TemporalContext> temporalContextObjectResponse = validateTemporalContext(geneExpressionFmsDTO);
 		if (temporalContextObjectResponse.hasErrors()) {
-			response.addErrorMessage("whenExpressed", temporalContextObjectResponse.errorMessagesString());
+			response.addErrorMessage("expressionPattern", temporalContextObjectResponse.errorMessagesString());
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java:116]: This change also doesn't make sense to me

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 3]
```diff
@@ -238,29 +224,29 @@ private ObjectResponse<AnatomicalSite> validateAnatomicalSite(GeneExpressionFmsD
 
 			if (!ObjectUtils.isEmpty(geneExpressionFmsDTO.getWhereExpressed().getAnatomicalStructureQualifierTermId())) {
 				String anatomicalstructurequalifiertermId = geneExpressionFmsDTO.getWhereExpressed().getAnatomicalStructureQualifierTermId();
-				VocabularyTerm anatomicalStructureQualifierTerm = vocabularyTermService.getTermInVocabulary(VocabularyConstants.ANATOMICAL_STRUCTURE_UBERON_SLIM_TERMS, anatomicalstructurequalifiertermId).getEntity();
+				OntologyTerm anatomicalStructureQualifierTerm = ontologyTermService.findByCurieOrSecondaryId(anatomicalstructurequalifiertermId);
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: You need to validate both that it is a valid ontology term (which you are doing) and that it comes from the appropriate VocabularyTermSet (which you are not doing)

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 4]
```diff
@@ -238,29 +224,29 @@ private ObjectResponse<AnatomicalSite> validateAnatomicalSite(GeneExpressionFmsD
...
 				if (anatomicalStructureQualifierTerm == null) {
-					response.addErrorMessage("whereExpressed - anatomicalStructureQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + geneExpressionFmsDTO.getWhereExpressed().getAnatomicalStructureTermId() + ")");
+					response.addErrorMessage("whereExpressed - anatomicalStructureQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + anatomicalstructurequalifiertermId + ")");
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java:229]: As above, you need to validate both that it is a valid ontology term (which you are doing) and that it comes from the appropriate VocabularyTermSet (which you are not doing)

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 5]
```diff
@@ -238,29 +224,29 @@ private ObjectResponse<AnatomicalSite> validateAnatomicalSite(GeneExpressionFmsD
...
 				}
 			}
 
 			if (!ObjectUtils.isEmpty(geneExpressionFmsDTO.getWhereExpressed().getAnatomicalSubStructureQualifierTermId())) {
 				String anatomicalsubstructurequalifierId = geneExpressionFmsDTO.getWhereExpressed().getAnatomicalSubStructureQualifierTermId();
-				VocabularyTerm anatomicalSubStructureQualifierTerm = vocabularyTermService.getTermInVocabulary(VocabularyConstants.ANATOMICAL_STRUCTURE_UBERON_SLIM_TERMS, anatomicalsubstructurequalifierId).getEntity();
+				OntologyTerm anatomicalSubStructureQualifierTerm = ontologyTermService.findByCurieOrSecondaryId(anatomicalsubstructurequalifierId);
 				if (anatomicalSubStructureQualifierTerm == null) {
-					response.addErrorMessage("whereExpressed - anatomicalSubStructureQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + geneExpressionFmsDTO.getWhereExpressed().getAnatomicalSubStructureTermId() + ")");
+					response.addErrorMessage("whereExpressed - anatomicalSubStructureQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + anatomicalsubstructurequalifierId + ")");
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: As above, you need to validate both that it is a valid ontology term (which you are doing) and that it comes from the appropriate VocabularyTermSet (which you are not doing)

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 6]
```diff
@@ -238,29 +224,29 @@ private ObjectResponse<AnatomicalSite> validateAnatomicalSite(GeneExpressionFmsD
...
 				}
 			}
 
 			if (!ObjectUtils.isEmpty(geneExpressionFmsDTO.getWhereExpressed().getCellularComponentQualifierTermId())) {
 				String cellularComponentQualifierTermId = geneExpressionFmsDTO.getWhereExpressed().getCellularComponentQualifierTermId();
-				VocabularyTerm cellularComponentQualifierTerm = vocabularyTermService.getTermInVocabulary(VocabularyConstants.CELLULAR_COMPONENT_QUALIFIERS, cellularComponentQualifierTermId).getEntity();
+				OntologyTerm cellularComponentQualifierTerm = ontologyTermService.findByCurieOrSecondaryId(cellularComponentQualifierTermId);
 				if (cellularComponentQualifierTerm == null) {
-					response.addErrorMessage("whereExpressed - cellularComponentQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + geneExpressionFmsDTO.getWhereExpressed().getCellularComponentQualifierTermId() + ")");
+					response.addErrorMessage("whereExpressed - cellularComponentQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + cellularComponentQualifierTermId + ")");
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: As above, you need to validate both that it is a valid ontology term (which you are doing) and that it comes from the appropriate VocabularyTermSet (which you are not doing)

**Comment by markquintontulloch** [comment]
I'm guessing you also need to update the integration tests to use ontology terms instead of vocabulary terms

File: src/test/java/org/alliancegenome/curation_api/ExpressionBulkUploadFmsITCase.java [hunk 7]
```diff
@@ -159,18 +167,24 @@ private void loadRequiredEntities() throws Exception {
 		createVocabularyTerm(vocabulary2, VocabularyConstants.GENE_EXPRESSION_RELATION_TERM, false);
 		Vocabulary stageUberonTermVocabulary = getVocabulary(VocabularyConstants.STAGE_UBERON_SLIM_TERMS);
 		Vocabulary anatomicalStructureUberonTermVocabulary = getVocabulary(VocabularyConstants.ANATOMICAL_STRUCTURE_UBERON_SLIM_TERMS);
-		Vocabulary cellullarComponentQualififerVocabulary = getVocabulary(VocabularyConstants.CELLULAR_COMPONENT_QUALIFIERS);
+		Vocabulary spatialExpressionQualififerVocabulary = getVocabulary(VocabularyConstants.SPATIAL_EXPRESSION_QUALIFIERS);
 		createStageTerm(stageTermId, "StageTermTest");
 		createVocabularyTerm(stageUberonTermVocabulary, stageUberonTermId, false);
 		createAnatomicalTerm(anatomicalStructureTermId, "AnatomicalStructureTermTest");
-		createAnatomicalTerm(anatomicalSubstructureTermId, "AnatomicalSubStructureTermTest");
+        createAnatomicalTerm(anatomicalSubstructureTermId, "AnatomicalSubStructureTermTest");
```
Comment (devopsagr) [inline src/test/java/org/alliancegenome/curation_api/ExpressionBulkUploadFmsITCase.java:174]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineJavaCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Indent must use tab characters

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java [hunk 8]
```diff
@@ -238,31 +224,37 @@ private ObjectResponse<AnatomicalSite> validateAnatomicalSite(GeneExpressionFmsD
 
 			if (!ObjectUtils.isEmpty(geneExpressionFmsDTO.getWhereExpressed().getAnatomicalStructureQualifierTermId())) {
 				String anatomicalstructurequalifiertermId = geneExpressionFmsDTO.getWhereExpressed().getAnatomicalStructureQualifierTermId();
-				VocabularyTerm anatomicalStructureQualifierTerm = vocabularyTermService.getTermInVocabulary(VocabularyConstants.ANATOMICAL_STRUCTURE_UBERON_SLIM_TERMS, anatomicalstructurequalifiertermId).getEntity();
-				if (anatomicalStructureQualifierTerm == null) {
-					response.addErrorMessage("whereExpressed - anatomicalStructureQualifierTermId", ValidationConstants.INVALID_MESSAGE + " (" + geneExpressionFmsDTO.getWhereExpressed().getAnatomicalStructureTermId() + ")");
-				} else {
-					anatomicalSite.setAnatomicalStructureQualifiers(List.of(anatomicalStructureQualifierTerm));
+				if (vocabularyTermService.getTermInVocabulary(VocabularyConstants.SPATIAL_EXPRESSION_QUALIFIERS, anatomicalstructurequalifiertermId) != null) {
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: These checks need to be validating that the term is in the appropriate VocabularyTermSet, not the whole Vocabulary

**Review by markquintontulloch — CHANGES_REQUESTED** [review]

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java (hunk 1 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java:107]: With that change the errors will be like "expressionPattern - whereExpressed - error message"

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java (hunk 2 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java:116]: With that change the errors will be like "expressionPattern - whenExpressed - error message"

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java (hunk 3 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: Done

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java (hunk 5 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: Done

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java (hunk 6 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: Done

File: src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java (hunk 8 above)
Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: Done

Comment (abecerra) [reply src/main/java/org/alliancegenome/curation_api/services/validation/dto/fms/GeneExpressionAnnotationFmsDTOValidator.java]: Done

**Comment by abecerra** [comment]
> The AnatomicalSite entity class needs the `@AGRCurationSchemaVersion` annotation updated to a min of 2.7.0 - you will also need to increase the `LinkMLSchemaConstants.LATEST_RELEASE` constant to this value

Done

File: src/main/java/org/alliancegenome/curation_api/constants/VocabularyConstants.java [hunk 9]
```diff
@@ -69,14 +69,17 @@ private VocabularyConstants() {
...
-	
+
 	public static final String TRANSCRIPT_RELATION_VOCABULARY = "transcript_relation";
 	public static final String TRANSCRIPT_CHILD_TERM = "is_child_of";
 	public static final String TRANSCRIPT_PARENT_TERM = "is_parent_of";
 
 	public static final String STAGE_UBERON_SLIM_TERMS = "stage_uberon_slim_terms";
 	public static final String ANATOMICAL_STRUCTURE_UBERON_SLIM_TERMS = "anatomical_structure_uberon_slim_terms";
-	public static final String CELLULAR_COMPONENT_QUALIFIERS = "cellular_component_qualifiers";
+	public static final String SPATIAL_EXPRESSION_QUALIFIERS = "spatial_expression_qualifiers";
```
Comment (markquintontulloch) [inline src/main/java/org/alliancegenome/curation_api/constants/VocabularyConstants.java:79]: Can we add _VOCABULARY or _VOCABULARY_TERM_SET suffixes to these names to make it clear what these constants refer to

**Review by markquintontulloch — APPROVED** [review]

**Pull request merged**